tasks/04_pages/task.py

```
@@ -0,0 +1,79 @@
+
+def main():
+    query = """
+            SELECT t1.id, t1.case_number
+            FROM cases t1
+                LEFT JOIN pages t2 ON t1.id = t2.case_id
+                WHERE t2.case_id IS NULL;
+            """
+
+    try:
+        with sql.db_cnx() as cnx:
+            c = cnx.cursor()
+            print('Fetching cases left to scrape.')
+            c.execute(query)
+            cases = [x for x in c.fetchall()]
+    except Exception as e:
+        print(f'Unable to gather remaining cases.')
+        raise(e)
+    else: # no exception
+        if len(cases) == 0:
+            print('Scrape complete')
+            return
+        else:
+            print('...remaining to scrape:', len(cases))
+    finally:
+        c.close()
+        cnx.close()
+    
+
+    # run the scraper. can tune the `max_workers` for multi-threading.
+    # note that each thread/worker opens and closes a connection to the db.
+    try:
+        with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
+            executor.map(scraper.threaded_scraper, cases)
+
+    # if manually ending scraper, update and display the number of cases remaining
+    except KeyboardInterrupt:
+        print('Scrape stopped!')
+        executor.shutdown(cancel_futures=True, wait=False)
+    
+    else:
+        print("Scrape completed")
+
+    finally:
+        try:
+            with sql.db_cnx() as cnx:
+                print('Attempting to count remaining cases and clean empty rows...')
+                c = cnx.cursor()
+                count_query = 'select (select count(*) from cases) - (select count(*) from pages) as row_diff;'
+                c.execute(count_query)
+                num_cases = c.fetchone()[0]
+                c.execute(
+                    """
+                    DELETE FROM pages WHERE raw_text = '';
+                    """
+                    )
+        except Exception as e:
+            print(f'Unable to count cases or clean empty rows: {e}')
```

For clarity, maybe this could be "Unable to count remaining case pages to scrape or clean empty rows"

---

You probably also want to raise this exception instead of just printing it, since in the future we may have a top-level script that coordinates running all the separate tasks, and any exceptions should be bubbled up to its level to be handled there

---

that makes sense!